src/main/java/org/rumbledb/items/parsing/ItemParser.java

```diff
@@ -62,13 +63,45 @@
     private static final DataType vectorType = new VectorUDT();
     public static final DataType decimalType = new DecimalType(30, 15); // 30 and 15 are arbitrary
 
-    public static Item getItemFromObject(JsonIterator object, ExceptionMetadata metadata) {
+    public static Item getItemFromObject(JsonReader<Object> object, ExceptionMetadata metadata, boolean isTopLevel) {
         try {
-            if (object.whatIsNext().equals(ValueType.STRING)) {
-                return ItemFactory.getInstance().createStringItem(object.readString());
+            if (isTopLevel) {
+                object.getNextToken();
             }
-            if (object.whatIsNext().equals(ValueType.NUMBER)) {
-                String number = object.readNumberAsString();
+            byte token = object.last();
```

Comment:
The API offered by dslJson is an "interesting" one. While getNextToken() and last() do the job, they are not very descriptive and can even be misleading for new developers (ie. last() -> gets the mostRecentByte or the last byte in the stream?)

I would be highly in favor of defining our own interface to dslJson which abstract away these details and provide a cleaner API. It would also make our code much more resilient and easy to fix in case something changes in dslJson's API in the future. Or even more, in case we decide to change the supporting library dslJson with something else.

What do you think?

Edit: On a second thought, we are parsing the JSON input character by character with various checks at each turn. I wonder if we are losing the speed benefit against GSON in this process. Not to mention readability. I would be very curious to see the performance comparison of these two implementations.

Reply:
Many thanks for your thorough comments.

Regarding the dslJson API: I agree -- the low-level is actually a plus though, because we can support more parsing options (lenient, duplicates, etc) that we couldn't with gson or jsonIter.

Regarding adding an API: we could consider adding our own internal API -- since this is a critical portion of the code that gets executed billions of times though, we need to be careful if we add more interfaces and method calls than there is already, and carefully check the performance impact if we do so.

Reply:
>Regarding the dslJson API: I agree -- the low-level is actually a plus though, because we can support more parsing options (lenient, duplicates, etc) that we couldn't with gson or jsonIter.

This is beautiful, I had not considered this.

> Regarding adding an API: we could consider adding our own internal API -- since this is a critical portion of the code that gets executed billions of times though, we need to be careful if we add more interfaces and method calls than there is already, and carefully check the performance impact if we do so.

Fair point, I may try to refactor this part in a separate branch after we get the initial implementation working. Performance reports can be generated thereafter